### PR TITLE
Implemented KAS repository to load transaction data

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Added `Receipts` field to `ChainEvent`.
- Added repository interface with KAS repository implementation.
- This is imported from `kas-app-server`, but transforming data is different.
- Introduced request and response structs.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		// to
+		to := rawTx.To()
+		if to == nil {
+			to = &common.Address{}
+		}
+
+		// value
+		value := hexutil.EncodeBig(rawTx.Value())
+
+		// transaction hash
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository_transactions.go:64]: no process?

File: datasync/chaindatafetcher/types.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package chaindatafetcher
+
+import (
+	"github.com/klaytn/klaytn/blockchain"
+	"math/big"
+)
+
+type requestType int
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/types.go:8]:
```suggestion
type requestType uint
```

File: datasync/chaindatafetcher/kas/repository.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+func getEndpoint(user, password, host, port, name string) string {
+	return user + ":" + password + "@tcp(" + host + ":" + port + ")/" + name + "?parseTime=True&charset=utf8mb4"
+}
+
+func NewRepository(user, password, host, port, name string) *repository {
+	endpoint := getEndpoint(user, password, host, port, name)
+	db, err := gorm.Open("mysql", endpoint)
+	if err != nil {
+		logger.Crit("Connecting to DB is failed", "endpoint", endpoint, "err", err)
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository.go:33]: How about try again with error log?

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 4]
```diff
@@ -125,7 +139,12 @@ func (f *ChainDataFetcher) resLoop() {
 			logger.Info("stopped resLoop for chaindatafetcher")
 			return
 		case res := <-f.resCh:
-			f.updateCheckpoint(res)
+			if res.err != nil {
+				logger.Error("db insertion is failed", "blockNumber", res.blockNumber, "err", res.err)
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/chaindata_fetcher.go:143]:
```suggestion
				logger.Error("db insertion is failed", "blockNumber", res.blockNumber, "reqType", res.reqType, "err", res.err)
```

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	types.TxTypeFeeDelegatedSmartContractDeploy:          true,
+	types.TxTypeFeeDelegatedSmartContractDeployWithRatio: true,
+
+	types.TxTypeSmartContractExecution:                      true,
+	types.TxTypeFeeDelegatedSmartContractExecution:          true,
+	types.TxTypeFeeDelegatedSmartContractExecutionWithRatio: true,
+}
+
+// transformToKasTxs transforms a chain event to transaction list according to the KAS database scheme.
+func transformToKasTxs(event blockchain.ChainEvent) []*Tx {
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository_transactions.go:34]: This code is in `kas` repository. I think we don't need to add `kas` something.

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	var txs []*Tx
+	// TODO-ChainDataFetcher extract EOAs in order to expire the KAS cache
+	//expiredEOAs := make(map[string]struct{})
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository_transactions.go:37]:
```suggestion
	//expiredEOAs := make(map[common.address]struct{})
```

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Applied ae3595c96d5d49ef7411c28543d63200048f17a1

File: datasync/chaindatafetcher/kas/repository.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Reconnect logic is added on f1f349c67257ed5a4fb0833cfcf62805d65744a2

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go]: Right. I renamed the methods as well. 6c5074d74c98fafee0346dbc327c37015c5e4913

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go]: No process. Removed the comment d0b0c6cb753b422717a9f7388278c6d73a38fe17

File: datasync/chaindatafetcher/types.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/types.go]: Applied. 95dbf1b8514529d09ea897bb5c6ed8f5048dbede

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go]: Applied. 11356affb7ef62522cf5e9b03b0989796b4d51ef

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
+package kas
```
[the remark this answers is not in the document]
Comment (ehnuje) [reply datasync/chaindatafetcher/kas/repository_transactions.go:1]: Are you planning to add more modules other than `repository`?

File: datasync/chaindatafetcher/types.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+const (
+	requestTypeTransaction = requestType(1) << iota
+)
+
+type request struct {
+	reqType requestType
+	event   blockchain.ChainEvent
+}
+
+type response struct {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/types.go:22]: Added on 2daf89797b3ce258b5d7133db5ec0c9247900836.

File: datasync/chaindatafetcher/repository.go [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package chaindatafetcher
+
+import "github.com/klaytn/klaytn/blockchain"
+
+type Repository interface {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/repository.go]: Not really. I made it private on 9f57bb12e4f13ac5bf4881b817d74eb18f8a8235.

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go:1]: Models for KAS database and repository methods will be in the package. I think it is more clear if any other package is added such as `kinesis`. Any thought?

File: datasync/chaindatafetcher/kas/repository.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+func NewRepository(user, password, host, port, name string) (*repository, error) {
+	endpoint := getEndpoint(user, password, host, port, name)
+	var (
+		db  *gorm.DB
+		err error
+	)
+	for i := 0; i < maxDBRetryCount; i++ {
+		db, err = gorm.Open("mysql", endpoint)
+		if err != nil {
+			logger.Info("Retrying to connect DB", "endpoint", endpoint, "err", err)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kas/repository.go:42]: Shouldn't this be `Warn`?

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+		}
+
+		txs = append(txs, tx)
+	}
+	return txs
+}
+
+// InsertTransactions inserts transactions in the given chain event into KAS database.
+func (r *repository) InsertTransactions(event blockchain.ChainEvent) error {
+	return r.insertTransactions(transformToTxs(event))
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kas/repository_transactions.go:114]: It seems that `insertTransactions` can be directly used here, doesn't it?

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ehnuje) [reply datasync/chaindatafetcher/kas/repository_transactions.go:1]: Nope! I just wondered.

File: datasync/chaindatafetcher/kas/repository.go (hunk 10 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Right. Chainged it 9fa1c7acb42f8f23587d26b76a1976565d98b074.

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 11 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go]: Right. Applied on 8b05945844acaebf5f4581850db0bf10f66460bf.

File: datasync/chaindatafetcher/kas/repository.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jinzhu/gorm"
+	"github.com/klaytn/klaytn/log"
+	"time"
+)
+
+const (
+	maxTransactionCount         = int64(1000000)
+	maxTransactionLogCount      = int64(100000)
+	maxInternalTransactionCount = int64(10000)
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kas/repository.go:13]: Does `maxInternalTransactionCount` mean literally the maximum internal transaction count? If it does, why this value is smaller than `maxTransactionCount`? I think `maxTransactionCount` should be the smallest value among them.

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	var txs []*Tx
+	// TODO-ChainDataFetcher extract EOAs in order to expire the KAS cache
+	//expiredEOAs := make(map[common.Address]struct{})
+
+	block := event.Block
+	head := block.Header()
+	receipts := event.Receipts
+
+	for idx, rawTx := range block.Transactions() {
+		txId := head.Number.Int64()*maxTransactionCount*maxTransactionLogCount + int64(idx)*maxInternalTransactionCount
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kas/repository_transactions.go:44]: I don't get this equation. Where does it from?

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+			from, _ = types.Sender(signer, rawTx)
+		} else {
+			from, _ = rawTx.From()
+		}
+
+		// to
+		to := rawTx.To()
+		if to == nil {
+			to = &common.Address{}
+		}
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kas/repository_transactions.go:59]: Is it okay if `to` value is not updated to the generated contract address?

File: datasync/chaindatafetcher/kas/repository_transactions.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+		if rawTx.IsFeeDelegatedTransaction() {
+			payer, _ := rawTx.FeePayer()
+			feePayer = payer.Bytes()
+		}
+
+		ratio, ok := rawTx.FeeRatio()
+		if ok {
+			feeRatio = uint(ratio)
+		}
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kas/repository_transactions.go:87]:
```suggestion

			ratio, ok := rawTx.FeeRatio()
			if ok {
				feeRatio = uint(ratio)
			}
		}
```

File: datasync/chaindatafetcher/kas/repository.go (hunk 12 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go:13]: These values are predefined in KAS for unique id.

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 13 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go:44]: This is a transaction id for uniqueness.

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 14 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go:59]: Yes, we don't store the contract address in `klay_transfers` table.

File: datasync/chaindatafetcher/kas/repository_transactions.go (hunk 15 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository_transactions.go]: Applied on 1c220b0

File: datasync/chaindatafetcher/kas/repository.go [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jinzhu/gorm"
+	"github.com/klaytn/klaytn/log"
+	"time"
+)
+
+const (
+	maxTransactionCount         = int64(1000000)
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository.go:11]:
```suggestion
	maxTransactionCountPerBlock         = int64(1000000)
```

File: datasync/chaindatafetcher/kas/repository.go [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+	maxTransactionLogCount      = int64(100000)
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository.go:12]:
```suggestion
	maxTransactionLogCountPerTransaction     = int64(100000)
```

File: datasync/chaindatafetcher/kas/repository.go (hunk 12 above)
Comment (KimKyungup) [inline datasync/chaindatafetcher/kas/repository.go:13]:
```suggestion
	maxInternalTransactionCountPerTransaction = int64(10000)
```

File: datasync/chaindatafetcher/kas/repository.go (hunk 16 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Renamed constantas ca1645925c2cf450d54e6f8a59eb9cb58bd7f285

File: datasync/chaindatafetcher/kas/repository.go (hunk 17 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Renamed constantas ca1645925c2cf450d54e6f8a59eb9cb58bd7f285.

File: datasync/chaindatafetcher/kas/repository.go (hunk 12 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Renamed constantas ca1645925c2cf450d54e6f8a59eb9cb58bd7f285.

**Pull request merged**